### PR TITLE
Repair step list and execution from occ

### PR DESCRIPTION
**Pull request by tomneedham**
`occ maintenance:repair --list` and `occ maintenance:repair --single "OC\Repair\MoveAvatarOutsideHome"`

Question: Should we force enable maintenance mode?

### CODE REVIEW

File: core/Command/Maintenance/Repair.php [hunk 1]
```diff
@@ -63,6 +64,18 @@ protected function configure() {
 		$this
 			->setName('maintenance:repair')
 			->setDescription('Repair the installation.')
+			->addOption(
+				'list',
+				null,
+				InputOption::VALUE_NONE,
+				'Lists all possible repair steps'
```
Comment (PVince81) [inline core/Command/Maintenance/Repair.php]: "possible" or the ones that will run during upgrade ? there are "before upgrade" and "during upgrade" steps

**Review by PVince81 — COMMENTED** [review]
Cool thing. Expecting codecov complains. Needs tests.

File: core/Command/Maintenance/Repair.php (hunk 1 above)
Comment (tomneedham) [reply core/Command/Maintenance/Repair.php]: Hmm I only grabbed the ones from \OC\Repair::getRepairSteps and the other expensive ones. Are they just the core ones? Do I need to parse the appinfos for steps as well?

Comment (PVince81) [reply core/Command/Maintenance/Repair.php]: Apps can also provide repair steps so might be good to parse these as well, yes.

I'm not aware of any apps using this currently though.

File: core/Command/Maintenance/Repair.php [hunk 2]
```diff
@@ -116,6 +173,10 @@ protected function execute(InputInterface $input, OutputInterface $output) {
 		$this->config->setSystemValue('maintenance', $maintenanceMode);
 	}
 
+	private function runRepair() {
+
+	}
```
Comment (tomneedham) [inline core/Command/Maintenance/Repair.php]: woops

**Comment by PVince81** [comment]
abandoned PR ? moving to triage for rescheduling

**Comment by PVince81** [comment]
@tomneedham do you intend to finish this some day or should we take over ?

**Comment by PVince81** [comment]
I'd say:
- [ ] only let the command run if the server is in maintenance mode.

don't force maintenance mode as it can have unforeseen consequences and might surprise an admin (and all users of the instance in case it takes longer...)

**Comment by PVince81** [comment]
@tomneedham I think most of the rest is straightforward.

@karakayasemi please take over and finish the PR.

**Comment by karakayasemi** [comment]
- the repair steps of applications added to list option (the `dav` app has repair steps). 
- only `--list` option run if the server is not in maintenance mode.
- I refactored the code a bit and added the tests.

@tomneedham, @PVince81 please review. I will squash my commit after review.

**Comment by karakayasemi** [comment]
@ownclouders rebase

**Comment by ownclouders** [comment]
Hey! I'm [GitMate.io](https://gitmate.io)! This pull request is being rebased automatically. Please **DO NOT** push while rebase is in progress or your changes would be lost permanently :warning:

**Comment by karakayasemi** [comment]
unrelated UI test fail. Rebased to try one more time.

**Comment by ownclouders** [comment]
Automated rebase with [GitMate.io](https://gitmate.io) was successful! :tada:

File: core/Command/Maintenance/Repair.php [hunk 3]
```diff
@@ -144,4 +177,49 @@ public function handleRepairFeedBack($event) {
 				break;
 		}
 	}
+
+	/**
+	 * @param OutputInterface $output
+	 * @@return IRepairStep[]
+	 */
+	public function getAppsRepairSteps(OutputInterface $output) {
+		$apps = \OC::$server->getAppManager()->getInstalledApps();
```
Comment (VicDeo) [inline core/Command/Maintenance/Repair.php]: inject AppManager and use `$this->appManager->getEnabledAppsForUser()` instead. We can get rid of L190-192 then.

File: core/Command/Maintenance/Repair.php [hunk 4]
```diff
@@ -144,4 +177,49 @@ public function handleRepairFeedBack($event) {
...
+		$steps = [];
+		foreach ($apps as $app) {
+			try {
+				if (!\OC_App::isEnabled($app)) {
+					continue;
+				}
+				$info = \OC_App::getAppInfo($app);
```
Comment (VicDeo) [inline core/Command/Maintenance/Repair.php]: `$this->appManager->getAppInfo($app);`

**Comment by karakayasemi** [comment]
@VicDeo all issues are addressed. Please review again.

**Review by VicDeo — APPROVED** [review]
Let's wait for CI

**Comment by VicDeo** [comment]
@PVince81 backport?

**Comment by PVince81** [comment]
@karakayasemi please backport

**Comment by phil-davis** [comment]
Backport `stable10` #34499

**Comment by pmaier1** [comment]
Has this been documented?

**Comment by karakayasemi** [comment]
Could not find any ticket for documentation. Raised in here: https://github.com/owncloud/docs/issues/946

**Pull request merged**